File: interview_exercises/lru_cache.py
class DoublyLinkedList:
    def __init__(self, data=None):
        self.data = data
        self.next = None
        self.prev = None


class LruCache:
    def __init__(self, capacity=256):
        self.head = None
        self.tail = None
        self.capacity = capacity
        self.hash_size = 10000
        self.cache = [None] * self.hash_size

    def get(self, data_pos):
        data_ix = data_pos - 1
        data_ix %= self.hash_size

        node = self.cache[data_ix]
        if node is not None:
            data = node.data[1]
        else:
            data = -1
        print('GET {0}->{1}'.format(data_pos, data))
        return data

    def set(self, data_pos, data):
        data_ix = data_pos - 1
        data_ix %= self.hash_size

        new_head = self.cache[data_ix]
        if new_head is not None:
            new_head.data = (data_ix, data)
            prev_node = new_head.prev
            next_node = new_head.next
            if prev_node is not None:
                prev_node.next = next_node
            next_node.prev = prev_node
        else:
            new_head = DoublyLinkedList((data_ix, data))
            if self.head is None:
                self.head = new_head
                self.tail = new_head
            if self.capacity > 0:
                self.capacity -= 1
            else:
                self.cache[self.tail.data[0]] = None
                new_tail = self.tail.next
                new_tail.prev = None
                self.tail.next = None
                self.tail = new_tail

        self.head.next = new_head
        self.head = new_head
        self.cache[data_ix] = new_head
        print('SET {0}->{1}'.format(data_pos, data))


if __name__ == '__main__':
    lru = LruCache(capacity=2)

    lru.set(1, 2)

    lru.set(2, 3)
    lru.set(1, 5)
    lru.set(4, 5)
    lru.set(6, 7)
    lru.get(4)

    lru.get(1)

File: cpp/cpp11/lrucache.cpp
#include <iostream>
#include <memory>
#include <unordered_map>
#include <list>
#include <vector>

using namespace std;

template<typename X, typename T>
class LruCache {
private:
    struct ListNode {
        explicit ListNode(pair<X, T> data) : data(move(data)) {};

        ~ListNode() { cout << data.first << ": destroyed\n"; };
        pair<X, T> data;
        shared_ptr<struct ListNode> prev;
        shared_ptr<struct ListNode> next;
    };

    shared_ptr<struct ListNode> head;
    shared_ptr<struct ListNode> tail;

    unordered_map<X, shared_ptr<struct ListNode>> lruMap;
    T capacity = 4;
public:
    LruCache() = default;

    ~LruCache() = default;

    T get(X ix);

    void set(pair<X, T> keyValue);

    void printMap();

    void printList();
};

template<typename X, typename T>
void LruCache<X, T>::printMap() {
    for (auto p : lruMap) {
        cout << p.first << " : " << p.second->data.second << "\n";
    }
    cout << "\n";
}

template<typename X, typename T>
void LruCache<X, T>::printList() {
    auto next_node = head;
    while (next_node) {
        auto data = next_node->data;
        cout << "key: " << data.first << ", value: " << data.second << "\n";
        next_node = next_node->next;
    }
    cout << "\n";
}

template<typename X, typename T>
T LruCache<X, T>::get(X ix) {
    auto lru_it = lruMap.find(ix);
    X value = -1;
    if (lru_it != lruMap.end()) {
        auto newHead = lru_it->second;
        value = newHead->data.second;
        auto next = newHead->next;
        auto prev = newHead->prev;

        if (prev) {
            prev->next = next;
        }
        if (next) {
            next->prev = prev;
        }

        newHead->prev = nullptr;
        newHead->next = head;
        head->prev = newHead;
        head = newHead;
    }
    return value;
}

template<typename X, typename T>
void LruCache<X, T>::set(pair<X, T> keyValue) {
    auto oldValue = get(keyValue.first);
    if (oldValue >= 0) {
        head->data.second = keyValue.second;
    } else {
        auto newHead = make_shared<struct ListNode>(keyValue);
        lruMap[keyValue.first] = newHead;
        if (head == nullptr) {
            head = newHead;
            tail = head;
        } else {
            newHead->next = head;
            head->prev = newHead;
            head = newHead;
        }
    }

    if (capacity > 0) {
        --capacity;
    } else {
        /* remove lru, i.e. tail and from map*/
        lruMap.erase(tail->data.first);

        auto newTail = tail->prev;
        newTail->next = nullptr;
        tail = newTail;
    }
}

int32_t main(int32_t argc, char **argv) {
    vector<double> v{12., 0., 2., 100., 1., 24.};
    auto lruCache = LruCache<int32_t, double>{};
    for (auto i = 0; i < v.size(); ++i) {
        lruCache.set(pair<int32_t, double>{i, v[i]});
    }

    lruCache.printMap();

    for (auto key = 0; key < v.size(); ++key) {
        auto value = lruCache.get(key);
        cout << "key: " << key << ", value: " << value << "\n";
    }
    cout << "\n";

    lruCache.printList();

    lruCache.get(4);

    lruCache.printList();

}
